When BUILD_DOCS is enabled, allow choosing if it should be in default make rule & install doc index.html

jmarrec opened the pull request:
Pull request overview
---------------------
 - Fixes #7315 again following #7826
 - When building from E+ and not standalone: if `BUILD_DOCS` is enabled, add a new option `BUILD_DOCS_ONLY_WITH_PACKAGE`:
   * If OFF (default): add docs target to default build rule
   * If ON: add a `docs` target but not to default build rule. But call it when `package` is called


cf https://github.com/NREL/EnergyPlus/pull/7826#issuecomment-612956581

- https://github.com/NREL/EnergyPlus/pull/7923/commits/16ced9bf8e13f349aaecfe7b6cb639285f405fec fixes #7953 and correctly installs index.html. Tested on both Windows and Ubuntu with the QtIFW installer.  Cf comment below https://github.com/NREL/EnergyPlus/pull/7923#issuecomment-626822596

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmarrec committed:
When BUILD_DOCS is enabled, add a `doc` target but not to default. But call it when `package` is called

cf https://github.com/NREL/EnergyPlus/pull/7826#issuecomment-612956581

jmarrec committed:
Add a companion flag `BUILD_DOCS_ONLY_WITH_PACKAGE`

jmarrec commented on CMakeLists.txt:
```diff
@@ -248,6 +248,8 @@ if( BUILD_PACKAGE )
 endif()
 
 if (BUILD_DOCS)
+  option(BUILD_DOCS_ONLY_WITH_PACKAGE "Build DOCS only explicitly via `make docs` or when `make package`" OFF)
+  # mark_as_advanced(BUILD_DOCS_ONLY_WITH_PACKAGE)
```
New option. Wasn't sure if it should be made 'advanced' or not, I choose not to (so people know it's a thing)

nealkruis replied:
I think it should be advanced.

Myoldmopar replied:
I also think it should be advanced, but I agree that since we are defaulting to OFF, it may be OK to leave it as non-advanced for a while, then advance it for a later cleanup phase.

jmarrec commented on doc/CMakeLists.txt:
```diff
@@ -21,8 +21,17 @@ if( EXISTS ${XELATEX} )
 
-  # Add custom target to build only documentation
-  add_custom_target(docs ALL)
+  # Add custom target to build only documentation, this target will depend on all individual PDFs
+  get_directory_property(hasParent PARENT_DIRECTORY)
+  if(hasParent AND BUILD_DOCS_ONLY_WITH_PACKAGE)
+    # Because we don't want to rebuild doc automatically, we do not add it to the 'ALL' rule
+    message(AUTHOR_WARNING "The target `docs` was added, but not the default build rule. It will be triggered by `make docs` or `make package`")
+    add_custom_target(docs)
+  else()
+    # Building standalone, so add to ALL
+    message(AUTHOR_WARNING "The target `docs` was added to the default build rule")
+    add_custom_target(docs ALL)
+  endif()
```
Handle whether it's built standalone or as part of E+. If E+, handle the `BUILD_DOCS_ONLY_WITH_PACKAGE`

jmarrec commented on cmake/Install.cmake:
```diff
@@ -489,6 +489,49 @@ set(CPACK_PROJECT_CONFIG_FILE "${PROJECT_BINARY_DIR}/CMakeCPackOptions.cmake")
+
+    # Call the build of target docs explicitly here.
+    # Note: This is because you can't do `add_dependencies(package docs)` (https://gitlab.kitware.com/cmake/cmake/issues/8438)
+    # Adding another custom target to be added to the "ALL" one (so it runs) and make it depend on the actual "documentation" target doesn't work
+    # because it'll always run if you have enabled BUILD_DOCS, regardless of whether you are calling the target "package" or not
+    #  add_custom_target(run_documentation ALL)
+    #  add_dependencies(run_documentation documentation)
+    #message(FATAL_ERROR "CMAKE_COMMAND=${CMAKE_COMMAND}")
+
+    # +env will pass the current environment and will end up respecting the -j parameter
+    #                                 this ↓↓↓ here -- https://stackoverflow.com/a/41268443/531179
+    #install(CODE "execute_process(COMMAND +env \"${CMAKE_COMMAND}\" --build \"${PROJECT_BINARY_DIR}\" --target documentation)")
+    # Except it doesn't work with install(execute_process...
+
+    # Passing $(MAKE) doesn't work either, and isn't a great idea for cross platform support anyways
+    # install(CODE "execute_process(COMMAND ${MAKE} ${DOC_BUILD_FLAGS} -C \"${PROJECT_BINARY_DIR}\" documentation)")
+
+    # So instead, we just used the number of threads that are available. That's not ideal, since it ignores any "-j N" option passed by the user
+    # But LaTeX should run quickly enough to not be a major inconvenience.
+    # There no need to do that for Ninja for eg, so only do it for Make and MSVC
+
+    # flag -j to cmake --build was added at 3.12 (VERSION_GREATER_EQUAL need cmake >= 3.7, we apparently support 2.8...)
+    if(NOT(CMAKE_VERSION VERSION_LESS "3.12") AND ((CMAKE_GENERATOR MATCHES "Make") OR WIN32))
+      include(ProcessorCount)
+      ProcessorCount(N)
+      if(NOT N EQUAL 0)
+        set(DOC_BUILD_FLAGS "-j ${N}")
+      endif()
+    endif()
+    if(WIN32)
+      # Win32 is multi config, so you must specify a config when calling cmake.
+      # Let's just use Release, it won't have any effect on LaTeX anyways.
+      set(DOC_CONFIG_FLAG "--config Release")
+    endif()
+
+    # Getting these commands to work (especially with macro expansion) is tricky. Check the resulting `cmake_install.cmake` file in your build folder if need to debug this
+    install(CODE "execute_process(COMMAND \"${CMAKE_COMMAND}\" --build \"${PROJECT_BINARY_DIR}\" ${DOC_CONFIG_FLAG} ${DOC_BUILD_FLAGS} --target docs)"
+            COMPONENT Documentation)
+  endif()
+
```
If `BUILD_DOCS_ONLY_WITH_PACKAGE`, then explicitly call `cmake --build --target docs`

nealkruis replied:
This seems pretty hacky. I would think there would be a way to add a dependency to the install target, but [apparently not](https://stackoverflow.com/questions/9163176/cmake-add-depedency-to-the-install-target). If it works for you then I don't see a problem with it.

jmarrec committed:
Fix #7953 - install doc index.html

jmarrec commented:
https://github.com/NREL/EnergyPlus/pull/7923/commits/16ced9bf8e13f349aaecfe7b6cb639285f405fec fixes #7953 and correctly installs index.html. Tested on both Windows and Ubuntu with the QtIFW installer.

Also tested that the Windows Startup menu entry works. 

![image](https://user-images.githubusercontent.com/5479063/81588174-6b332180-93b8-11ea-925b-e42c8df0f46b.png)

![image](https://user-images.githubusercontent.com/5479063/81588223-800fb500-93b8-11ea-8dbf-caa3b14a8ff5.png)


That allowed me to test the earlier changes again on both ubuntu and windows, works fine, when `BUILD_DOCS_ONLY_WITH_PACKAGE `is ON, I get the docs built with package but not the default (`ALL_BUILD`) target. When `BUILD_DOCS_ONLY_WITH_PACKAGE `is OFF, docs are always built.

I can separate this into another PR if you want @Myoldmopar

jmarrec committed:
Typo in author warning.

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into Build_docs_adjustments

nealkruis commented on doc/CMakeLists.txt:
```diff
@@ -21,8 +21,17 @@ if( EXISTS ${XELATEX} )
+    add_custom_target(docs)
+  else()
+    # Building standalone, so add to ALL
+    message(AUTHOR_WARNING "The target `docs` was added to the default build rule")
```
I don't think there needs to be a special message for the default case.

jmarrec replied:
I would have made `BUILD_DOCS_ONLY_WITH_PACKAGE=ON` by default myself.
And therefore, since it isn't, both making it advanced and not displaying a message for like one iteration cycle at least is likely going to mean that many developpers might not notice this option at all.

That being said, if there is consensus that this is better to make it advanced and not display a message I'm totally fine with it and it's a very quick change.


----

As to your comment, above (there's no reply box):

 > This seems pretty hacky. I would think there would be a way to add a dependency to the install target, but apparently not. If it works for you then I don't see a problem with it.

I would have thought there would be a way too, but quickly realized there wasn't.

Myoldmopar replied:
I'd like to see these warnings go away.  At least the one for the default case.  These are going to pop up in my console every time I configure with BUILD_DOCS on.

My real preference would be to eliminate the `BUILD_DOCS` and `BUILD_DOCS_ONLY_WITH_PACKAGE` variables, and have one single variable that defines the documentation build config, with options instead of just a boolean status: `DOCUMENTATION_BUILD: [ DoNotBuild | BuildWithAll | BuildOnlyWithPackage ]` or something like that.  But I think that has too many possible ramifications.  

I'm OK with leaving things as they are, but not throwing a warning.

nealkruis commented on cmake/Install.cmake:
```diff
@@ -503,9 +546,10 @@ if ( BUILD_DOCS )
   install(FILES "${PROJECT_BINARY_DIR}/doc/pdf/PlantApplicationGuide.pdf" DESTINATION "./Documentation" COMPONENT Documentation)
   install(FILES "${PROJECT_BINARY_DIR}/doc/pdf/TipsAndTricksUsingEnergyPlus.pdf" DESTINATION "./Documentation" COMPONENT Documentation)
   install(FILES "${PROJECT_BINARY_DIR}/doc/pdf/UsingEnergyPlusForCompliance.pdf" DESTINATION "./Documentation" COMPONENT Documentation)
+  install(FILES "${PROJECT_BINARY_DIR}/doc/pdf/index.html" DESTINATION "./Documentation" COMPONENT Documentation)
```
Is this fixing something I did (did we ever package this file)?

mjwitte replied:
@nealkruis No, not something you did. Fixing  https://github.com/NREL/EnergyPlus/issues/7953 (should be noted at the top).

nealkruis reviewed (commented):
Just a couple of comments to discuss here.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into Build_docs_adjustments

Myoldmopar reviewed (changes requested):
I'd like to get rid of the warning message for the default case.  Other than that, I'm OK with this as it is.

jmarrec committed:
Merge remote-tracking branch 'origin/develop' into Build_docs_adjustments

jmarrec committed:
Address review comments: make BUILD_DOCS_ONLY_WITH_PACKAGE advanced, no message in the default case, and just a status for the other.

jmarrec committed:
Add Multi-choice option instead of two booleans + backward compat

Temporary workaround to default the correct DOCUMENTATION_BUILD value if BUILD_DOCS existed in the cache or is passed via the command line

jmarrec commented:
@Myoldmopar 6c9c08457238ee1f53e3fbb023a26c1c56338c00 is the minimal changes you asked for.

b1e7ef29a2f54032ef6958e3e1b556a098515f9f is an optional variant that does use the `DOCUMENTATION_BUILD: [ DoNotBuild | BuildWithAll | BuildOnlyWithPackage ]` multi choice option, with a (hopefully temporary) workaround preserving backwards compatibility with `BUILD_DOCS` (either from the Cache, or if passed explicitly via command line such as for a clean build).

mitchute commented:
I'm getting CMake errors here. What did I miss?

```
-- Could NOT find LATEX (missing: LATEX_COMPILER) 
CMake Error at doc/CMakeLists.txt:56 (message):
  Expected to find xelatex at: PDFLATEX_COMPILER-NOTFOUND ; check your LaTeX
  installation
```

```
mmitchel$ xelatex --version
XeTeX 3.14159265-2.6-0.999991 (TeX Live 2019)
kpathsea version 6.3.1
Copyright 2019 SIL International, Jonathan Kew and Khaled Hosny.
There is NO warranty.  Redistribution of this software is
covered by the terms of both the XeTeX copyright and
the Lesser GNU General Public License.
For more information about these matters, see the file
named COPYING and the XeTeX source.
Primary author of XeTeX: Jonathan Kew.
Compiled with ICU version 63.1; using 63.1
Compiled with zlib version 1.2.11; using 1.2.11
Compiled with FreeType2 version 2.9.1; using 2.9.1
Compiled with Graphite2 version 1.3.13; using 1.3.13
Compiled with HarfBuzz version 2.3.1; using 2.3.1
Compiled with libpng version 1.6.36; using 1.6.36
Compiled with poppler version 0.68.0
Using Mac OS X Core Text and Cocoa frameworks

mmitchel$ pdflatex --version
pdfTeX 3.14159265-2.6-1.40.20 (TeX Live 2019)
kpathsea version 6.3.1
Copyright 2019 Han The Thanh (pdfTeX) et al.
There is NO warranty.  Redistribution of this software is
covered by the terms of both the pdfTeX copyright and
the Lesser GNU General Public License.
For more information about these matters, see the file
named COPYING and the pdfTeX source.
Primary author of pdfTeX: Han The Thanh (pdfTeX) et al.
Compiled with libpng 1.6.36; using libpng 1.6.36
Compiled with zlib 1.2.11; using zlib 1.2.11
Compiled with xpdf version 4.01
```

nealkruis commented:
@mitchute I have seen this before on Mac where CMake isn't finding the LaTeX distribution. You can edit the advanced settings in CMake to explicitly point to the LaTeX binaries.

mitchute commented:
Thanks, @nealkruis. It looks like CLion is having some troubles finding the LaTeX package here and on develop. That doesn't look like a problem with this work. It all works fine when I run it from the terminal. 

I tested with the ```BuildWithAll``` and ```BuildOnlyWithPackage``` options, and it looks good. Any other comments before this goes in?

mitchute commented:
This won't cause any problems with the recent GIO merge to develop. Merging now.
